tools/extra/fpgadiag/nlb0.cpp

```diff
@@ -310,15 +314,43 @@ bool nlb0::setup()
     options_.get_value<bool>("suppress-hdr", suppress_header_);
     options_.get_value<bool>("csv", csv_format_);
 
+    options_.get_value<std::string>("id", nlb0_id_);
+    fpga_guid nlb0_id;
+
+    if (uuid_parse(nlb0_id_.c_str(), nlb0_id)<0) {
+        std::cerr << "Invalid NLB0 id:" << nlb0_id_ << std::endl;
+        return false;
+    }
+
+    uint32_t offset = 0;
+    uint64_t dfh = 0;
+    uint8_t* n = (uint8_t *)&nlb0_id;
+    uint64_t nlb0_hi = swap64(uint64_t, n);
+    uint64_t nlb0_lo = swap64(uint64_t, n+8);
+
+    do {
+        uint64_t feature_uuid_lo, feature_uuid_hi;
+        // Read the next feature header
+        dfh = accelerator_->read_csr64(static_cast<uint32_t>(offset));
+        feature_uuid_lo = accelerator_->read_csr64(static_cast<uint32_t>(offset+8));
+        feature_uuid_hi = accelerator_->read_csr64(static_cast<uint32_t>(offset+16));
+        if ((nlb0_lo == feature_uuid_lo)&&(nlb0_hi == feature_uuid_hi)) {
+            offset_ = offset;
+            printf("found the NLB offset=0x%x\n", offset);
+            break;
+        }
+        offset += NEXT_DFH_OFFSET(dfh);
+    } while (!DFH_EOL(dfh));
```

Comment:
If this terminates without setting offset_, then 348 reads from the wrong location and 409+ attempts to talk to the wrong device.

Reply:
offset_  default value is zero.

ORG Code 
 // set dsm base, high then low
    accelerator_->write_csr64(static_cast<uint32_t>(nlb0_dsm::basel), reinterpret_cast<uint64_t>(dsm_->io_address()));

Modified By VC team

 accelerator_->write_csr64(static_cast<uint32_t>(nlb0_dsm::basel)+offset_, reinterpret_cast<uint64_t>(dsm_->io_address()));

they added offeset_ to code , default value is zero


![image](https://user-images.githubusercontent.com/33236856/57410146-e27b1780-719e-11e9-8834-f68239d76164.png)

Reply:
I understand that it was added.  The errors are still present and should not go into public master.